File: backend/src/infra/code-search/git-history.ts
import { simpleGit } from "simple-git";
import { getErrorMessage } from "../../utils/error.js";
import { logger } from "../../utils/logger.js";

/**
 * Interface yang merepresentasikan informasi Git Blame per baris kode.
 */
export interface BlameLine {
    /** Nomor baris di dalam file */
    lineNumber: number
    /** Hash komit yang mengubah baris ini */
    hash: string
    /** Penulis/pengubah baris kode ini */
    author: string
    /** Tanggal komit dalam format ISO string */
    date: string
    /** Isi teks dari baris kode tersebut */
    content: string
}

/**
 * Interface yang merepresentasikan entri entri riwayat komit Git.
 */
export interface LogEntry {
    /** Hash komit (SHA-1) */
    hash: string
    /** Penulis komit */
    author: string
    /** Tanggal komit */
    date: string
    /** Pesan komit (commit message) */
    message: string
}

/**
 * Membaca riwayat commit (`git log`) untuk file tertentu di dalam repositori.
 * Digunakan oleh AI Agent untuk mencari tahu kapan dan mengapa sebuah file/baris berubah.
 * 
 * @param repoPath - Path direktori repositori lokal di disk
 * @param filePath - Path file relatif di dalam repositori
 * @param maxCount - Jumlah maksimal komit yang ingin diambil (default: 20)
 * @returns Array dari objek `LogEntry`
 */
export async function getFileLog(repoPath: string, filePath: string, maxCount = 20): Promise<LogEntry[]> {
    const git = simpleGit(repoPath);

    try {
        const log = await git.log({ file: filePath, maxCount })
        return log.all.map((entry) => ({
            hash: entry.hash,
            author: entry.author_name,
            date: entry.date,
            message: entry.message,
        }))
    } catch (error) {
        throw new Error(`Failed to read git log for ${filePath} in ${repoPath}: ${getErrorMessage(error)}`)
    }
}

/**
 * Membaca riwayat per baris kode (`git blame`) untuk file tertentu di dalam repositori.
 * Menggunakan perintah `git blame --line-porcelain` untuk mendapatkan format machine-readable.
 * 
 * @param repoPath - Path direktori repositori lokal di disk
 * @param filePath - Path file relatif di dalam repositori
 * @returns Array dari objek `BlameLine`
 */
export async function getFileBlame(repoPath: string, filePath: string): Promise<BlameLine[]> {
    const git = simpleGit(repoPath)
    try {
        // "--" wajib ada biar filePath (bisa dari argumen tool/untrusted) yang kebetulan
        // diawali "-" gak kebaca sebagai flag git, tapi tetep sebagai path.
        const raw = await git.raw(["blame", "--line-porcelain", "--", filePath])
        return parseBlamePorcelain(raw)
    } catch (error) {
        throw new Error(`Failed to blame ${filePath} in ${repoPath}: ${getErrorMessage(error)}`)
    }
}

/**
 * Mem-parsing output mentah dari perintah `git blame --line-porcelain` menjadi array dari `BlameLine`.
 * 
 * Format baris `--line-porcelain`:
 * - `<40-char-hash> <orig-line> <final-line>` (Header baris)
 * - `author <Nama Penulis>`
 * - `author-time <Epoch-seconds>`
 * - `\t<Isi Baris Kode>` (Diawali karakter Tab)
 * 
 * @param raw - Output string mentah dari git blame --line-porcelain
 * @returns Array dari `BlameLine`
 */
function parseBlamePorcelain(raw: string): BlameLine[] {
    const result: BlameLine[] = []
    let current: { hash?: string; author?: string; date?: string } = {}

    let lineNumber = 0

    for (const line of raw.split("\n")) {
        // RegEx untuk mencocokkan header 40 karakter heksadesimal SHA-1 hash dan nomor baris akhir
        const header = /^([0-9a-f]{40}) \d+ (\d+)/.exec(line)

        if (header) {
            logger.info({ hash: header[1], lineNum: header[2] }, "HEADER BLAME MATCH!")
            current = { hash: header[1] }
            lineNumber = Number(header[2])
            continue
        }

        if (line.startsWith("author ")) {
            current.author = line.slice("author ".length)
        } else if (line.startsWith("author-time ")) {
            const epochSeconds = Number(line.slice("author-time ".length))

            // Mengonversi epoch detik menjadi milidetik dan membentuk ISO String
            current.date = new Date(epochSeconds * 1000).toISOString()

        } else if (line.startsWith("\t")) {
            result.push({
                lineNumber,
                hash: current.hash ?? "",
                author: current.author ?? "",
                date: current.date ?? "",
                content: line.slice(1),
            })
        }
    }
    return result
}

File: engine/app/infra/code_search/git_history.py
"""1:1 port (behavior, bukan wire format internal) dari backend/src/infra/code-search/git-history.ts."""

import re
from dataclasses import dataclass
from datetime import UTC, datetime

from app.infra.proc import run_subprocess

_TIMEOUT_S = 30
_MAX_OUTPUT_BYTES = 10 * 1024 * 1024

# Separator ASCII (unit/record separator) - bukan reproduksi marker internal simple-git,
# cukup pastiin field & flag git command-nya sama (hash, author, ISO date, message, --follow).
_FIELD_SEP = "\x1f"
_RECORD_SEP = "\x1e"

_HEADER_RE = re.compile(r"^([0-9a-f]{40}) \d+ (\d+)")


@dataclass(frozen=True)
class LogEntry:
    hash: str
    author: str
    date: str
    message: str


@dataclass(frozen=True)
class BlameLine:
    line_number: int
    hash: str
    author: str
    date: str
    content: str


async def get_file_log(repo_path: str, file_path: str, max_count: int = 20) -> list[LogEntry]:
    result = await run_subprocess(
        [
            "git", "-C", repo_path, "log", f"--max-count={max_count}", "--follow",
            f"--pretty=format:%H{_FIELD_SEP}%aN{_FIELD_SEP}%aI{_FIELD_SEP}%s{_RECORD_SEP}",
            "--", file_path,
        ],
        cwd=None, timeout_s=_TIMEOUT_S, max_output_bytes=_MAX_OUTPUT_BYTES,
    )
    if not result.passed:
        raise RuntimeError(f"Failed to read git log for {file_path} in {repo_path}: {result.output}")

    entries: list[LogEntry] = []
    for record in result.output.split(_RECORD_SEP):
        record = record.strip("\n")
        if not record:
            continue
        h, author, date, message = record.split(_FIELD_SEP)
        entries.append(LogEntry(hash=h, author=author, date=date, message=message))
    return entries


async def get_file_blame(repo_path: str, file_path: str) -> list[BlameLine]:
    # "--" wajib ada biar file_path (bisa dari argumen tool/untrusted) yang kebetulan diawali "-"
    # gak kebaca sebagai flag git.
    result = await run_subprocess(
        ["git", "-C", repo_path, "blame", "--line-porcelain", "--", file_path],
        cwd=None, timeout_s=_TIMEOUT_S, max_output_bytes=_MAX_OUTPUT_BYTES,
    )
    if not result.passed:
        raise RuntimeError(f"Failed to blame {file_path} in {repo_path}: {result.output}")
    return _parse_blame_porcelain(result.output)


def _parse_blame_porcelain(raw: str) -> list[BlameLine]:
    result: list[BlameLine] = []
    current_hash = ""
    current_author = ""
    current_date = ""
    line_number = 0

    for line in raw.split("\n"):
        header = _HEADER_RE.match(line)
        if header:
            current_hash = header.group(1)
            line_number = int(header.group(2))
            continue

        if line.startswith("author "):
            current_author = line[len("author "):]
        elif line.startswith("author-time "):
            epoch_seconds = int(line[len("author-time "):])
            current_date = datetime.fromtimestamp(epoch_seconds, tz=UTC).isoformat().replace("+00:00", "Z")
        elif line.startswith("\t"):
            result.append(BlameLine(
                line_number=line_number, hash=current_hash, author=current_author,
                date=current_date, content=line[1:],
            ))

    return result
